meson: 0.57.1 -> 0.60.3, gst-plugins-base 1.18.4 -> 1.18.5

Princemachiavelli opened the pull request:
###### Motivation for this change

Replaces #140545 and this can be looked at after the current ZHF cycle.

https://mesonbuild.com/Release-notes-for-0-60-0.html

Updated meson for compatibility with newer packages such as wlroots 0.16
Updated gst-plugins-base for compatibility with meson >= 0.58

I just stumbled upon the meson version issue when building gamescope since it requires wlroots 0.16. https://github.com/Plagman/gamescope

nixpkgs-review is in progress although it might end up being more than practical to run on a single machine. i.e it seems this will trigger a rebuild of >40,000 packages just looking at nixpkgs-review now. :/

I have built meson itself & gst-plugins-base without issue.

Closes #150486.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Princemachiavelli committed:
libnice: added graphviz build input

added graphviz build input for compatability with meson >=0.57.2 see PR meson#8641

cdunster commented:
Hey, I'm very new to Nix but I'm currently trying to move my dev environment over to use Nix packages. However, I need meson ~v0.59.1 which is how I came across this PR. I thought it would just be a matter of changing the version in the `buildPythonApplication` section but there is a lot more to it. Do you mind explaining why? Is it because other packages depend on meson? Just trying to learn more about Nix, thanks :slightly_smiling_face:.

Also, is there anything I can do to help with this PR?

jtojnar commented on pkgs/development/tools/build-managers/meson/boost-Do-not-add-system-paths-on-nix.patch:
```diff
@@ -1,26 +1,9 @@
          else:
              tmp = []  # type: T.List[Path]
--
--            # Homebrew
```
Note: removed in https://github.com/mesonbuild/meson/commit/3c3fa0a12c68da8966410ad9bacd1148dc8bd65d

Princemachiavelli replied:
Yea, I removed the hombrew lines since those were already removed in the upstream.

Princemachiavelli commented:
Sorry for the delay on this, I got distracted and there was an upstream change to meson that required updating one of the patches (I'll annotate that in case anyone wants to review that).

Also, there is a pretty big change in meson 0.60; unknown options passed to `meson configure` and `meson setup` are not fatal. Options that are intended to be deprecated now should be configured as such in the `meson_options.txt`.

Practically this means that a lot more projects are going to break because it's been a common problem that packages accumulate meson build options more than we remove old/non-existent options. This means that this change will help us clean up a lot of packages but it also is going to be more hydra failures to fix even if the process is straightforward.

Princemachiavelli committed:
systemd: removed unknown meson options

Princemachiavelli commented on pkgs/development/tools/build-managers/meson/gir-fallback-path.patch:
```diff
@@ -1,17 +1,24 @@
          if fatal_warnings:
              scan_command.append('--warn-error')
- 
-+        if len(set([girtarget.get_custom_install_dir() for girtarget in girtargets])) > 1:
```
Upstream added type hinting to the get_custom_install_dir function so now it is a optional type containing a tupple (str,bool). This change also changed the default/null case; previously the default return value would be [None] but now the default case is simple an empty list [].

If a python-guru knows how to handle a empty set within a list comprehension, we could certainly make this less verbose. Also if anyone has any preference on how the patch is written such as how the empty list check is done or anything like that, I'm all ears.

jtojnar replied:
We could use `[girtarget.get_custom_install_dir()[0] if girtarget.get_custom_install_dir() is not None for girtarget in girtargets]`

jtojnar replied:
We do not even need the square brackets as Python supports set comprehensions.

Princemachiavelli replied:
Thanks I forgot a clause could be added to the end of the list/set comprehension. I went with just `set(girtarget.get_custom_install_dir()[0] for girtarget in girtargets if girtarget.get_custom_install_dir())`.

Princemachiavelli commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -418,7 +418,7 @@ stdenv.mkDerivation {
     "-Doomd=${lib.boolToString withOomd}"
     "-Dpolkit=${lib.boolToString withPolkit}"
-    "-Dcryptsetup=${lib.boolToString withCryptsetup}"
+    "-Dlibcryptsetup=${lib.boolToString withCryptsetup}"
```
Example case where we had an incorrect/outdated name of a meson option.

alyssais replied:
Doesn't look like systemd ever had a "cryptsetup" option, only "libcryptsetup".

Princemachiavelli commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -456,7 +456,6 @@ stdenv.mkDerivation {
     "-Dsysvinit-path="
     "-Dsysvrcnd-path="
 
-    "-Dkill-path=${coreutils}/bin/kill"
```
Case where we had a non-existent build option. What is hard to know is whether we needed that build option in for compatibility on Nix. i.e. did systemd start hard coding the path to `kill` or did they implement `kill` inside of systemd?

alyssais replied:
[They implemented kill inside systemd.](https://github.com/systemd/systemd/commit/9a85778412fa3e3f8d4561064131ba69f3259b28)

Identified with `git log -pG kill-path` in the systemd repo.

SuperSandro2000 commented on pkgs/development/libraries/gstreamer/good/default.nix:
```diff
@@ -48,7 +48,7 @@ assert raspiCameraSupport -> (stdenv.isLinux && stdenv.isAarch64);
+  version = "1.18.5";
 
   outputs = [ "out" "dev" ];
 
```
Hash did not get an update?

Princemachiavelli replied:
Strange, not sure why it would build on my machine if the hash is wrong (perhaps if it's pre-fetched it bypasses some check?). I've updated the hash.

alyssais commented:
Could you separate the Meson and Gstreamer changes into separate commits, and give 23dbcfc0cf59260b51fd16e8cb905c27979d9a96 a Nixpkgs-style commit message please?

alyssais commented:
shared-mime-info fails to build with Meson ≥0.60.0.  There's a [fix](https://gitlab.freedesktop.org/xdg/shared-mime-info/-/merge_requests/162) in progress upstream, but it's quite a big change and difficult to backport, so I think we should wait and see what happens.  Last activity was only a couple of days ago anyway.

alyssais commented:
@vcunat could we maybe get a Hydra jobset for this?  There are a bunch of incompatible changes that are breaking packages, and it's going to be hard to find them all otherwise.

Princemachiavelli committed:
meson: 0.57.1 -> 0.60.1

gir-fallback-path.patch: meson 0.60 added types to the get_custom_install_dir and changed the default return type

Princemachiavelli committed:
gst-plugins-base 1.18.4 -> 1.18.5

vcunat commented:
Yes in principle, but it's a matter of priorities / time.  I know of two similar feature-jobsets that were paused now due to Hydra being overloaded a bit.  (It's quite common at periods around forking a release.)

Mindavi commented:
I cherry-picked these commits on top of nixpkgs, but cross-compiling e.g. iputils fails:

`nix build -f default.nix pkgsCross.aarch64-multiplatform.iputils`

<details><summary>Error log</summary>

```
error: builder for '/nix/store/qcnmvz869vdc887237ganpkpcklgyznw-iputils-aarch64-unknown-linux-gnu-20210722.drv' failed with exit code 1;
       last 10 log lines:
       >   File "/nix/store/kmq06b06fxrspgk24q9wz038pqwyghd6-meson-0.60.1/lib/python3.9/site-packages/mesonbuild/mesonlib/universal.py", line 1393, in Popen_safe
       >     p = subprocess.Popen(args, universal_newlines=True, close_fds=False,
       >   File "/nix/store/nqck1zi4rj8sggdwwh1y21dkq3k624fh-python3-3.9.6/lib/python3.9/subprocess.py", line 951, in __init__
       >     self._execute_child(args, executable, preexec_fn, close_fds,
       >   File "/nix/store/nqck1zi4rj8sggdwwh1y21dkq3k624fh-python3-3.9.6/lib/python3.9/subprocess.py", line 1821, in _execute_child
       >     raise child_exception_type(errno_num, err_msg, err_filename)
       > OSError: [Errno 8] Exec format error: 'ldconfig'
       > FAILED: meson-install
       > /nix/store/kmq06b06fxrspgk24q9wz038pqwyghd6-meson-0.60.1/bin/meson install --no-rebuild
       > ninja: build stopped: subcommand failed.
```

</details>

I don't know if this is a regression in meson or in the packaging, but this will potentially break quite some cross-compilable packages as-is. I don't really have a list of other things I tried, but I tried some other packages on another day and those resulted in similar errors.

Princemachiavelli commented:
> I cherry-picked these commits on top of nixpkgs, but cross-compiling e.g. iputils fails:
> 
> `nix build -f default.nix pkgsCross.aarch64-multiplatform.iputils`

@Mindavi  Strange, I was able to build the same cross-compile of iputils on my meson-0.60 branch.

Given that the error show in the log you posted mentions an issue executing ldconfig:

` OSError: [Errno 8] Exec format error: 'ldconfig'`

I think the issue is the build host used does not have the `boot.binfmt.emulatedSystems` setting configured. In my case I have it set to contain "aarch64-linux". I would assume is required for cross compiling packages. Was it possible to cross-compile packages without configuring binfmt prior to this meson update?

Mindavi commented:
Yes. Emulation is not cross-compilation. Binfmt should not be required for cross-compilation.

Without using this change, iputils cross-compiles fine, so there is a regression somewhere.

Princemachiavelli committed:
meson: 0.60.1 -> 0.60.2

Princemachiavelli commented:
https://github.com/mesonbuild/meson/commit/2c3a68bc30cf6e08c2e3b77b64718ef3a0e5d78c

Looks like meson added a step to run ldconfig to update the ld.so.cache cache. It looks like the return code of the subprocess which runs ldconfig shouldn't matter but I guess the Nix build process is somehow observing the error occur.

I also just updated the PR to reflect the release of meson 0.60.2 . I don't see any commits between 0.60.1 and 0.60.2 that address this issue but it can't hurt to try to replicate the issue on 0.60.2

Mindavi commented:
It replicates on 0.60.2. This should probably be addressed upstream. `ldconfig` is not returning an error, Popen returns an exception because the cross-compiled ldconfig is (attempted to be) executed. That won't work.

Doing something like this works:

```
  postPatch = ''
    substituteInPlace mesonbuild/minstall.py \
      --replace "'ldconfig'" "'${buildPackages.glibc}/bin/ldconfig'"
  '';
```

But is not really a nice solution, I think. Besides, I think we don't care (don't use/build) about that cache at all. Maybe we want to be able to disable this 'cache build' action as a whole?

I also tried adding glibc to propagatedNativeBuildInputs (and removing the line that removes the propagation file), but that didn't help. Presumably the cross-compiled ldconfig comes earlier in the PATH.

Thanks for looking into this.

alyssais commented:
Thanks to everybody who's helped with this upgrade so far.

When QEMU 6.2.0 is released on either the 7th or the 14th, getting it in Nixpkgs will be blocked on this.

Princemachiavelli committed:
meson: disable updating ld.so cache

Princemachiavelli commented:
@Mindavi I just added a patch that simply removes the step that updates the ld.so cache. This should fix the problem and I don't think updating that cache has any advantage on Nix like you said. I'll try to find a moment to test it although I think all my machines have binfmt configured so it's a bit tricky.

@alyssais Thanks. Now that 20.11 has been released I think there should be some freed up time and hopefully Hydra resources for this.

FYI I found that using the Archlinux PKGBUILDs for meson related build changes was useful since Arch is already on meson 0.60.2 .

Mindavi commented:
I think that's fine for now. If you want to test you could use an architecture you're not using, but chances for a failure get higher the more exotic the architecture is.

Princemachiavelli commented:
I was able to temporarily disable binfmt via the /proc/sys/fs/binfmt_misc/status control file device. Building iputils for aarch64 worked without issue using the above patch.

alyssais committed:
shared-mime-info: 2.1 -> unstable-2021-12-03

We need https://gitlab.freedesktop.org/xdg/shared-mime-info/-/merge_requests/162
("migrate from itstool to msgfmt for creating translated XML") to fix
the build with Meson 0.60.0.

Signed-off-by: Patrick Hilhorst <git@hilhorst.be>

alyssais commented:
This patch fixes shared-mime-info, which has now merged Meson 0.60.0+ compatibility, but has not had a release yet:

```diff
From c4379778e9a5da01ef8e6f1fb4e6cda67753f202 Mon Sep 17 00:00:00 2001
From: Alyssa Ross <hi@alyssa.is>
Date: Fri, 3 Dec 2021 15:47:22 +0000
Subject: [PATCH] shared-mime-info: 2.1 -> unstable-2021-12-03

We need https://gitlab.freedesktop.org/xdg/shared-mime-info/-/merge_requests/162
("migrate from itstool to msgfmt for creating translated XML") to fix
the build with Meson 0.60.0.
---
 pkgs/data/misc/shared-mime-info/default.nix | 20 +++++---------------
 1 file changed, 5 insertions(+), 15 deletions(-)

diff --git a/pkgs/data/misc/shared-mime-info/default.nix b/pkgs/data/misc/shared-mime-info/default.nix
index 4c60895897f..2feb5e2f5a4 100644
--- a/pkgs/data/misc/shared-mime-info/default.nix
+++ b/pkgs/data/misc/shared-mime-info/default.nix
@@ -13,34 +13,24 @@
 }:
 
 stdenv.mkDerivation rec {
-  pname = "shared-mime-info";
-  version = "2.1";
+  pname = "shared-mime-info-unstable";
+  version = "2021-12-03";
 
   outputs = [ "out" "dev" ];
 
   src = fetchFromGitLab {
     domain = "gitlab.freedesktop.org";
     owner = "xdg";
-    repo = pname;
-    rev = version;
-    sha256 = "07bxv44p43pqq4ymfnyy50yli7lwdqymhvclna42rkn1cazq3vb5";
+    repo = "shared-mime-info";
+    rev = "5a406b06792e26a83c7346b3c2443c0bd8d4cdb2";
+    sha256 = "1v7dx7mr0m4lcff1aasg9gxn280zn0ffn6fjg9xc44pnllg01n6s";
   };
 
-  patches = [
-    # xmlto is only used for building the docs, which are not installed anyways.
-    (fetchpatch {
-      name = "xmlto-optional.patch";
-      url = "https://gitlab.freedesktop.org/xdg/shared-mime-info/-/merge_requests/110.patch";
-      sha256 = "0p5gxlcmn8ji5bc7pd105s1halqwa1d28lfx9yj43rn6mav7allx";
-    })
-  ];
-
   nativeBuildInputs = [
     meson
     ninja
     pkg-config
     gettext
-    itstool
     libxml2
   ] ++ lib.optional (stdenv.hostPlatform != stdenv.buildPlatform) shared-mime-info;
 
-- 
2.33.0
```

(@Princemachiavelli if you could enable the option to let committers push to your branch, it might make it easier to collaborate on a change like this that will probably require lots of package fixes.)

Princemachiavelli commented:
@alyssais I have the "Allow edits and access to secrets by maintainers" option checked. Is that not working?

Synthetica9 commented:
@alyssais I was just able to push after fetching the pr with `gh`, maybe pebcak?

dcbaker commented:
@Princemachiavelli I'm an upstream meson dev (and a nixos user) would you open an issue in the meson issue tracker about that? the ldconfig stuff has been an endless source of headaches.

dcbaker commented:
Also, feel free to mention me if you're having issues updating meson in nix, I'm happy to help resolve things

Princemachiavelli commented:
@dcbaker Thanks. I created an issue https://github.com/mesonbuild/meson/issues/9707.

I tried to provide distro agnostic information & example but I had issues following the docs on viewing individual build phases (https://nixos.wiki/wiki/Nixpkgs/Create_and_debug_packages) so I mostly went off the derivation build log. Hopefully what I put in the issue is useful.

dcbaker commented:
Yup, looks good to me.

Synthetica9 commented:
> it is not run at all when $DESTDIR is used, and nix is a package manager. Package managers traditionally use $DESTDIR because that's a fundamental part of building packages. So that should have sidestepped the problem, What is Nix doing here?

Couldn't we set $DESTDIR anyways, even if it's not strictly needed?

jtojnar commented:
We would have to set it to `/` and hope nothing breaks (I have seen programs that did not like paths starting with `//`, not sure how common that is). Also, if we set `DESTDIR` we will lose other hooks like `glib-compile-schemas` in many apps – we would need to re-add those manually or create a hook similar to what Debian has.

Princemachiavelli commented:
Upstream meson did fix the ldconfig during cross build issue in https://github.com/mesonbuild/meson/pull/9708. However, I think we should keep the patch that disables it entirely because it can sometimes be slow and provides no benefit on NixOS. Even if it only costs a few seconds that will still add up over the many packages Hydra builds.

Mindavi commented:
Yeah, hopefully it'll be an option in the future.

dcbaker commented:
Im still pushing to remove the feature for 0.61, as is Eli, who's an arch Linux maintainer. But yeah, that's a breaking change for a point release so it won't happen till 0.61

AndersonTorres commented:
If no one object, can I update Meson tomorrow?

Princemachiavelli commented:
Since #150486 added support for multiple meson versions, do we want to continue supporting that and have this PR just change the default? It could be helpful to have multiple meson versions; we might find some packages have upstream issues with meson 0.60+.

jtojnar commented:
I consider having multiple versions (especially when forking the expression) a maintenance nightmare. I would just revert the previous PR as a first step of this one (you can insert `break` command at the top when doing interactive rebase and revert the commits before continuing the rebase).

vcunat commented:
I'd hope that meson is keeping quite good backward compatibility.

Princemachiavelli committed:
Revert "meson: document the failed tests issue"

This reverts commit 650c249d8f641c79623b80b0789feca42d2a30d5.

Princemachiavelli committed:
Revert "meson: add myself as maintainer"

This reverts commit 5dc14c358c96fb2eb54fbea4f10e8ba7d1032fd4.

Princemachiavelli committed:
Revert "meson 0.60: Fix more-env-vars.patch"

This reverts commit c401c98393d3ca0ce29ad09b7b9ba11b3856e825.

Princemachiavelli committed:
Revert "meson: 0.57.1 -> 0.60.2"

This reverts commit fac4c247ca3f19400ddefce6a88d4c50b0a539e3.

Princemachiavelli committed:
Revert "meson: copy 0.57 to 0.60"

This reverts commit e9c8925d1bc83625489dea6ef446f8218f9b25f8.

Princemachiavelli committed:
Revert "meson: put 0.57 in a subdirectory"

This reverts commit 1a53eeda71a17d917f89e6539094795435b1458e.

alyssais commented:
<details><summary>There are still some references to meson_0_60 that need to be cleaned up.</summary>
<pre>pkgs/applications/misc/metadata-cleaner/default.nix:12:, meson_0_60
pkgs/applications/misc/metadata-cleaner/default.nix:38:    meson_0_60
pkgs/development/libraries/libadwaita/default.nix:8:, meson_0_60
pkgs/development/libraries/libadwaita/default.nix:42:    meson_0_60
pkgs/development/libraries/wlroots/0.15.nix:1:{ lib, stdenv, fetchFromGitLab, meson_0_60, ninja, pkg-config, wayland-scanner
pkgs/development/libraries/wlroots/0.15.nix:26:  nativeBuildInputs = [ meson_0_60 ninja pkg-config wayland-scanner ];
</pre></details>

AndersonTorres commented:
> Since #150486 added support for multiple meson versions, do we want to continue supporting that and have this PR just change the default?

The end goal is to get rid of old Meson.

The other PR is to insert meson 0.60.x without disrupting the entire package base with unseen incompatibilities or mass rebuilds.

The next steps should be:
- changing the default to newest version
- fix problematic packages
- remove old meson

jtojnar commented:
More `meson_0_60` dependents from #152948.

L-as commented:
Instead of this huge PR that updates a bunch of things at once, what about flipping around the `meson_0_60` and `meson` names to `meson_0_59` and `meson`?

L-as commented:
Seeing that this PR is almost done I instead made #154172 to try to fix the remaining issues.

Princemachiavelli committed:
meson: remove versioned meson references from PR#152948

Princemachiavelli committed:
meson: 0.60.2 -> 0.60.3

alyssais commented:
Merged as #154172 since it had a history cleanup.  Thanks @Princemachiavelli for all your hard work!
